components/service/sync-autofill/src/main/java/mozilla/components/service/sync/autofill/GeckoCreditCardsAddressesStorageDelegate.kt

```diff
@@ -38,6 +40,22 @@ class GeckoCreditCardsAddressesStorageDelegate(
     }
 
     override fun onCreditCardSave(creditCard: CreditCard) {
-        TODO("Not yet implemented")
+        val validationDelegate = DefaultCreditCardValidationDelegate(storage)
```

Comment:
If this method is called in a quick succession (and you're not in control of its consumers, so you don't know if it'll be called in this fashion or not), it's possible that one of the add/update calls to storage below will fail (and we'll crash) with a SQL_LOCKED exception coming out of storage.

That's because sqlite doesn't support consequent writers, and within `AutofillCreditCardsAddressesStorage` you only have a single connection instance to the storage that's shared for both reading and writing. And, the `coroutineContext` use there is defined over `Dispatchers.IO`. That means you call `addCreditCard` and then very quickly an `updateCreditCard` while first operation is still running, it's possible for things to crash - we'll end up using a different thread for the update operation, and will try to talk to the storage while it's busy.

For comparison, `PlacesStorage` defines two scopes - `readScope` and `writeScope`. The former uses a thread pool to make sure reads are quick, while the latter uses a single threaded executor to make sure write operations are sequential and never overlap.

Just to avoid this class of errors and not have to worry about what's possible here and what's not, this is the kind of split I'd introduce here as well.